server/datastore/cached_mysql/cached_mysql.go

```diff
@@ -2,51 +2,113 @@ package cached_mysql
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"reflect"
 	"time"
 
 	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jinzhu/copier"
 	"github.com/patrickmn/go-cache"
 )
 
+const (
+	appConfigKey                      = "AppConfig:%s"
+	defaultAppConfigExpiration        = 1 * time.Second
+	packsHostKey                      = "Packs:host:%d"
+	defaultPacksExpiration            = 1 * time.Minute
+	scheduledQueriesKey               = "ScheduledQueries:pack:%d"
+	defaultScheduledQueriesExpiration = 1 * time.Minute
+	teamAgentOptionsKey               = "TeamAgentOptions:team:%d"
+	defaultTeamAgentOptionsExpiration = 1 * time.Minute
+)
+
+// cloner represents any type that can clone itself. Used by types to provide a more efficient clone method.
+type cloner interface {
+	Clone() (interface{}, error)
+}
+
+func clone(v interface{}) (interface{}, error) {
+	if cloner, ok := v.(cloner); ok {
+		return cloner.Clone()
+	}
+
+	vv := reflect.ValueOf(v)
+	isPtr := false
+	if vv.Kind() == reflect.Ptr {
+		isPtr = true
+		vv = vv.Elem()
+	}
+
+	clone := reflect.New(vv.Type())
+	err := copier.Copy(clone.Interface(), v)
+	if err != nil {
+		return nil, err
+	}
+
+	if isPtr {
+		return clone.Interface(), nil
+	}
+
+	return clone.Elem().Interface(), nil
```

Comment:
Nice, the reflection comes out looking pretty clean. Can you please add a comment explaining what the reflection code is doing?

Reply:
Added in 382a5000